stream: add flatMapPrefix operator

eyalfa opened the pull request:
## implement #28364.

## Changes
introduce a new flow/source operator `prefixAndDownstream` (name subject to change) that consumes up to n elements from upstream, then obtains the downstream flow by applying a user provided function with the consumed prefix, the flow is than materialized and serves as the downstream of this flow/source. This operator is useful for handling stateful protocols, i.e. obtain the schema/metadata before the actual data arrives, or reading a header and then deciding how to parse the following chunk of data.


## Background Context
This operator somehow resembles the existing `prefixAndTail` but it brings two additions to it:
1. the ability to access the materialized value of the downstream flow. (`prefixAndTail` exposes the tail as a source that can be pipelined to a flow chosen/constructed based on the prefix, but it does not provide access to this flow's materialized value).
2. unlike `prefixAndTail` the downstream flow is always materialized, so implementation doesn't have to rely on timers in order to check if the downstream wasn't materialized.

eyalfa committed:
akka-28364: initial implementation of the new graph stage and operator.

eyalfa committed:
akka-28364: working (?) graph stage + couple of tests. docs not included yet.

eyalfa committed:
akka-28364: added java dsl + initial (copy pasted) docs.

eyalfa committed:
akka-28364: documentation.

eyalfa commented:
@johanandren I'd appreciate your comments, especially re. naming and docs.

johanandren commented on akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:
```diff
@@ -1929,6 +1929,31 @@ trait FlowOps[+Out, +Mat] {
+   *  @param n the number of elements to accumulate before materializing the downstream flow.
+   *  @param f a function that produces the downstream flow based on the upstream's prefix.
+   **/
+  def prefixAndDownstream[Out2, Mat2](n : Int)(f : immutable.Seq[Out] => Flow[Out, Out2, Mat2]): Repr[Out2] = {
```
I like the signature, and having a separate Mat version is good.

Some naming consideration though, I still think it should be a `flatten` since that is much what it does, `flatMapPrefixAndTail` is a mouthful but I think it best conveys what the operator does.

eyalfa replied:
I think this operator is somewhat of a dynamic version of via/viaMat.
notice there's no flattening here, we just connect two flows together when one of them is not known up-front.

having that said, I think you as part of the akka team should decide about this operator's name.

johanandren replied:
Yeah, it's not quite `flatMap` either, true, hmmm. I'll talk to the team and see if anyone else has a good idea.

johanandren replied:
@ennru and me had a discussion and came up the suggestion `fromPrefix` which good since it says exactly what it does: creates the flow from a prefix, not so nice inside a stream flow api definition though `source.map(yadi).fromPrefix(2)(prefix => flow)`. Perhaps `flowFromPrefix` to cover that case more nicely.

Enno also suggested something like `peek` since it peeks into the stream, but as such operations (queue for example) usually do not consume elements which this operator does.

johanandren replied:
I should add that making it something like `via` is not good name because that is not an operator so very different from this. This is somewhat alike `lazyFlow(() => Flow)` in that it delays creation (but further), that doesn't help much with naming ideas for me though.

johanandren replied:
Discussed this a bit back and forth with @jrudolph and I think we both ended up liking `flatMapPrefix` best. Seeing that both standalone and in a flow I think gives a good intuition for what it does if you have used any other monady APIs (like the Scala collections):

```scala
// flow api
source.map(yadi).flatMapPrefix(2)(prefix => flow).to(sink)

// standalone
val flow = Flow.flatMapPrefix(2)(prefix => flow)
```

It's not perfect, since in the flow api the left hand side value is a `Source`, and the flatten implies `A => C[A]`, however I think that is ok since such a signature wouldn't anyway make any sense and we can think of the `C` as stream instead of specifically `Source`.

johanandren commented on akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala:
```diff
@@ -133,6 +135,167 @@ import scala.util.control.NonFatal
+    FlowShape.of(Inlet[In]("in"), Outlet[Out]("out"))
+  }
+}
+
```
I know we have dumped all kinds of operators in two or three huge file previously but please put this operator in a separate file to keep things neat rather than pile up on that

eyalfa replied:
👍 will do

johanandren commented on akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala:
```diff
@@ -133,6 +135,167 @@ import scala.util.control.NonFatal
+   */
+  override val shape: FlowShape[In, Out] = {
+    FlowShape.of(Inlet[In]("in"), Outlet[Out]("out"))
+  }
```
Follow the style of defining the in and out as fields together with the shape up top in the class used in other stages, also name the in and out with the stage "StageName.in/out", that helps in debugging etc.

johanandren commented on akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala:
```diff
@@ -133,6 +135,167 @@ import scala.util.control.NonFatal
+
+      override def onUpstreamFinish(): Unit = {
+        if(null != subSource) {
+          subSource.complete()
```
I think a better style than checking if there is a value all over in the callbacks would be to have an initial In/OutHandler that is not initialized and then switch over to a separate handler when the sub stream has been materialized (we know it will never switch back to not being initialized, so no further checks needed then)

eyalfa replied:
we'd still need to check in postStop

johanandren commented on akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala:
```diff
@@ -133,6 +135,167 @@ import scala.util.control.NonFatal
+          case NonFatal(ex) =>
+            val neverMaterializedEx = new NeverMaterializedException(ex)
+            matPromise.failure(neverMaterializedEx)
+            throw neverMaterializedEx
```
This should rather be `failStage(ex)`, that it was never materialized is only for the matval-future

johanandren commented on akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala:
```diff
@@ -133,6 +135,167 @@ import scala.util.control.NonFatal
+            /**
+             * Called when the output port has received a pull, and therefore ready to emit an element, i.e. [[GraphStageLogic.push]]
+             * is now allowed to be called on this port.
+             */
```
remote scaladoc that came from the API

johanandren commented on akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala:
```diff
@@ -133,6 +135,167 @@ import scala.util.control.NonFatal
+          .to(subSink.sink)
+        val m = Try(interpreter.subFusingMaterializer.materialize(runnableGraph))
+        matPromise.complete(m)
+        m.get
```
I wonder if it is really important to be able to discern failure to materialize from failure to create the flow with the types, if we skip that this could be cleaner with a single `try  { create and materialize and set up subflows } catch { case ex => fail-promise-and-stage }`

eyalfa replied:
hmm, makes sense.
care must be taken to prevent 'dangling' sub-source or sub-sink situations

johanandren reviewed (commented):
Looks promising, will need quite a bit of test coverage.

eyalfa commented:
@johanandren , can you state some test coverage you'd like to see here?
I've included a test-suite testing some basic cases and another suite for testing reactive-streams compatibility, please lmk what else would you like to see?
I think we can add some property based tests testing behaviour for different input lengths, prefix lengths, downstream behaviour (take(n), ignore, cancelled, seq, fail after n...) upstream failures...
sky is the limit 😎

johanandren commented:
OK TO TEST

johanandren commented on akka-stream-tests-tck/src/test/scala/akka/stream/tck/PrefixAndDownstreamTest.scala:
```diff
@@ -0,0 +1,21 @@
+      .runWith(Sink.asPublisher(false))
+    publisher
+  }
+}
```
I don't think we need to add this to the TCK

johanandren commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowPrefixAndDownstreamSpec.scala:
```diff
@@ -0,0 +1,127 @@
+    }
+  }
+
+}
```
I must have missed this last pass, good that it's here.

I think it would be good to also cover (please also think and see if you can come up with any other corner case scenarios I have missed):

 * upstream completion without any elements
 * both downstream cancellation and failure 
   * without any elements coming downstream
   * as tight as possible with the prefix coming downstream (it will likely be impossible to control the order of events deterministically but all cases should work so still useful)
   * after nested flow has been materialized
 * abrupt termination (stopping the materializer)

eyalfa replied:
@johanandren what does 

> as tight as possible with the prefix coming downstream

mean? can you please point me to a test suite taking this approach?

johanandren replied:
As in doing the downstream cancellation immediately after sending the last prefix element (we don't have control over if the prefix element or the cancellation will reach the stage logic first so therefore it won't be deterministic)

eyalfa replied:
ok, I think I got it.
BTW, I'm in the process of adding tests using published and subscriber probes. the first case is: n=2, upstream completes immediately after subscriber requests elements (publisher produces no elements), this unveiled a slew of bugs in my code, will push soon.

does it matter if the publisher completes before or after demand? is it deterministic?

johanandren replied:
Waiting for demand would be a good way to know that at least the stage has pulled. If you complete right away you won't know if the stage did pull or not.

johanandren replied:
Some additional potential good things to cover that I could come up with:

 * already mentioned, but, materialization failing
 *  returned flow cancelling but continuing to emit elements, for example `take(0).concat(Source(1 to 4))`
 * returned flow completing downstream but continuing to consume elements, for example  `fromSinkAndSource(Sink.ignore, Source.empty)` 
 * downstream cancelation and failure, could be good using a `fromSinkAndSource(probe, probe)` as nested flow so that you can observe cancellation/failure coming from downstream and then propagate it
 * abrupt termination (stopping the materializer) see for example last test in SeqSinkSpec

eyalfa committed:
akka-28364: address some of the review comments.

eyalfa committed:
akka-28364: add tests, fix a major bug.

eyalfa committed:
Merge branch 'master' into akka-28364

eyalfa committed:
akka-28364: add some mor tests.

eyalfa committed:
akka-28364: formatting

eyalfa committed:
akka-28364: fix DslConsistencySpec test failures.

eyalfa committed:
akka-28364: scalafmtall

eyalfa commented:
@johanandren , I've added some tests and implemented some fixes. I'd like to hear from you re. more tests before I move on to implement the two refactoring actions you've requested (move to a separate file, two distinct handlers implementation for the two states we have).

johanandren commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowPrefixAndDownstreamSpec.scala:
```diff
@@ -0,0 +1,219 @@
+      suffixF.futureValue should ===(100 :: 101 :: Nil)
+    }
+
+    "handles upstream cancellation" in assertAllStagesStopped {
```
Upstream does not cancel, it completes. Not entirely sure this test adds anything on top of the `Source.empty` one above.

eyalfa replied:
mostly right, but this example shows that the flow may produce even when upstream produces no elements at all.

johanandren commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowPrefixAndDownstreamSpec.scala:
```diff
@@ -0,0 +1,219 @@
+
+      val matValue = Source
+        .fromPublisher(publisher)
+        .viaMat(if (true) flow else flow_)(Keep.right)
```
?

eyalfa replied:
debugging leftovers, tried figuring out if the problem was in the test or test-subject code.

eyalfa commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowPrefixAndDownstreamSpec.scala:
```diff
@@ -0,0 +1,219 @@
+      suffixF.futureValue /*(Interval(Span(1000, Minutes)))*/ should ===(10 until 20)
+    }
+
+    "propagate materialization failure" in assertAllStagesStopped {
```
1. will look at TE
2. you wrote in an earlier review we don't care wether the exception comes from 'f' application or flow's materialization...
3. will add a separate test for that.

eyalfa commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowPrefixAndDownstreamSpec.scala:
```diff
@@ -0,0 +1,219 @@
+    "propagate materialization failure" in assertAllStagesStopped {
+      val suffixF = Source(0 until 100)
+        .prefixAndDownstreamMat(10) { prefix =>
+          sys.error(s"I hate mondays! (${prefix.size})")
```
will do

eyalfa commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowPrefixAndDownstreamSpec.scala:
```diff
@@ -0,0 +1,219 @@
+        .to(Sink.ignore)
+        .run
+
+      val ex = suffixF.failed.futureValue
```
nice!

eyalfa commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowPrefixAndDownstreamSpec.scala:
```diff
@@ -0,0 +1,219 @@
+        .run()
+
+      prefixF.futureValue should be(empty)
+      suffixF.futureValue should be(empty)
```
materialized stream is not necessarily 'linear', hence it might produce multiple elements per input and (as you stated as a test case at the end of your review) it might concatenate with another stream/source. so upstream completion does not imply 'don't materialize the flow'

johanandren replied:
Good point

eyalfa committed:
akka-28364: more tests and fixes.

eyalfa committed:
akka-28364: scalafmt

eyalfa commented:
@johanandren , I'd appreciate if you could take another look, especially test-coverage.
is there any news about naming this operator?

raboof reviewed (commented):
Couple of comments, not fully reviewed yet.

As for naming, would it be too crazy to name this operator 'prefixAndTail` as well? The differences are mostly 'technical' rather than 'semantic', it's unambiguous (since they have a different number of parameters) and it makes the 2 variations easy to discover.

raboof commented:
> would it be too crazy to name this operator `prefixAndTail` as well

Ah I see this was considered at https://github.com/akka/akka/issues/28364#issuecomment-564952838 before.

eyalfa commented on akka-docs/src/main/paradox/stream/operators/Source-or-Flow/prefixAndDownstream.md:
```diff
@@ -0,0 +1,37 @@
+# prefixAndDownstream
+
+Take up to *n* elements from the stream (less than *n* only if the upstream completes before emitting *n* elements), then apply *f* on these elements in order to obtain a flow, this flow is then materialized and the rest of the input is processed by this flow (similar to via). This method returns a flow consuming the rest of the stream producing the materialized flow's output.
```
@raboof , that's the exact intent behind this operator! I think I'll change the short doc according to your suggestion and leave the elaboration in place for the longer doc.

eyalfa commented:
@raboof , the more I work on this PR the more I think of it as a dynamic variant of `via`/`viaMat`, so I really think the name should reflect that or perhaps some verb from of the `pipe` operation.
I'd also like to see more elaborate versions of both this and `prefixAndTail` that resembles `takeWhile`...

eyalfa committed:
akka-28364: address some more review comments, add tests for cases where n = 0.

eyalfa committed:
akka-28364: formatting for all

eyalfa committed:
akka-28364: yet another test case for n=0

eyalfa committed:
akka-28364: updated index.md

johanandren commented on akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala:
```diff
@@ -132,6 +134,152 @@ import scala.util.control.NonFatal
+          subSource = null
+          subSink = null
+          failStage(ex)
+        }
```
I still think this whole section would be cleaner with a try-catch doing the same stuff than the `Try` wrapping, it would also save 4 lambda allocs.

johanandren reviewed (commented):
Another round of review and confirmed with @raboof we agree on `flatMapPrefix` so please go ahead and change to that thanks!

eyalfa commented on akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala:
```diff
@@ -132,6 +134,152 @@ import scala.util.control.NonFatal
+              if (!isClosed(in)) {
+                cancel(in, cause)
+              }
+              //super.onDownstreamFinish(cause)
```
notice this is the sub-source's outlet, its downstream is the materialized flow, so this downstream-finished signal comes from the materialized flow - in case in is still open we propagate the signal upstream, if it's already closed we can't (I think I've introduced this condition after experiencing some test failures).

I'll revisit the existing tests and c if your suggested use-case isn't already covered.

johanandren replied:
I may have mixed it up. 

So if a cancellation comes from the inner flow, and upstream is completed already, the cancellation could either come from the inner flow itself, `fromSinkAndSource` for example, which has its downstream end still running, and in that case we should ignore the cancellation here (could be that we should act different if it is a propagated failure from downstream rather than just the normal cancel which will always be of type `akka.stream.SubscriptionWithCancelException.NonFailureCancellation`) and let the inner flow continue running, but it could also come from the actual downstream and have passed up through the inner flow, in that case out will be closed so we can check that as well.

I may be missing some additional complexity of this though.

eyalfa replied:
I don't think we should care too much about this:
1. if in is still open, we propagate the signal upstream
2. if in is already closed, the materialized flow has already processed this signal and probably decided to keep going and completed just now.
3. if out is already closed, then closing in indicates all stages have completed, hence the stage can be closed. next hing that happens is a call to onStop. M I missing anything?

can u think of a flow that closes both ends but keeps running? if such a case exists we might see the outer stage being closed (postStop called) while the inner stage is still running, I'm not sure what this is going to do...

johanandren replied:
Isn't the problem that since you swallow the cancellation here if in is closed, so the stage will never actually stop but keep running, or what triggers stopping it?

I think the problematic scenario would be:
1. stage consumes prefix, materializes inner flow
2. source completes
3. downstream cancels or fails
4. cancel or fail comes through inner flow to here
5. in is closed, inner flow is probably stopped but this stage keeps running (leak)

You should be able to detect such a scenario with probes and having the test complete, assertAllStagesStopped wrapping the test should notice the graph is still alive and fail.

eyalfa replied:
as demonstrated in test case 'run a detached flow',
when both in and out are closed the 'external' stage stops, in an extreme (made up?) scenario when the materialized flow keeps running, the external stage stops while the nested flow keeps running. when nested flow completes its stages are simply shutdown since inlets and outlets are being closed so active ports drops to zero for these stages and the interpreter simply closes them.

eyalfa committed:
akka-28364: address some of the review comments. add tests demonstrating early downstream cancellation/failure and the way it forces flow materialization.

johanandren commented on akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala:
```diff
@@ -132,6 +134,152 @@ import scala.util.control.NonFatal
 
   override def toString: String = s"FlattenMerge($breadth)"
 }
+@InternalApi private[akka] final class PrefixAndDownstream[In, Out, M](n: Int, f: immutable.Seq[In] => Flow[In, Out, M])
```
It's passed to the constructor so first thing here to fail as early as possible

eyalfa replied:
oh, ok.
but n=0 is a valid case

johanandren replied:
Hmm, is it really, we already have `lazyFlow` for that use case, and an empty prefix isn't really a prefix?

johanandren commented on akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala:
```diff
@@ -132,6 +134,152 @@ import scala.util.control.NonFatal
+    val logic = new GraphStageLogic(shape) with InHandler with OutHandler {
+      var accumulated = List.empty[In]
+      var subSource: SubSourceOutlet[In] = null
+      var subSink: SubSinkInlet[Out] = null
```
`akka.util.OptionVal` is better (but strictly internal) since it is a value class and compiles down to zero overhead.

eyalfa commented on akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala:
```diff
@@ -132,6 +134,152 @@ import scala.util.control.NonFatal
+      setHandlers(in, out, this)
+
+      override def postStop(): Unit = {
+        matPromise.tryFailure(new AbruptStageTerminationException(this))
```
AbruptStageTerminationException extends NoStackTrace so I guess it's not too expensive to instantiate... I'll clean this up anyway :-)

johanandren replied:
Ah, good, then you can decide yourself if you want that micro-optimization or not, won't be important.

eyalfa committed:
akka-28364: add test case for a nested flow that closes both ends but keeps running.

eyalfa committed:
akka-28364: scalafmt

eyalfa commented on akka-docs/src/main/paradox/stream/operators/Source-or-Flow/prefixAndDownstream.md:
```diff
@@ -0,0 +1,37 @@
+then apply *f* on these elements in order to obtain a flow, this flow is then materialized and the rest of the input is processed by this flow (similar to via).
+This method returns a flow consuming the rest of the stream producing the materialized flow's output.
+
+## Reactive Streams semantics
```
I do have a specific use case, but it's a bit to complex to just paste it in... it'll take me some time to simplify it into this format.
my use case: deserializing apache-arrow records from bytes-stream, serializer first writes the schema, then the data, deserializer first parses the schema then the data. parsing the data requires the schema as it describes what to be parsed: number of vectors (=columns), their names and data type. part of this information (names) would simply be redundant to write for every record (actually batch) and some is strictly required for proper parsing.

johanandren replied:
Yeah, that may be too much of a special case, I think there's two nice things we could show in an example, one is using the prefix to add to each element and the other is selecting the flow to use based in the prefix. But as I said you can leave that out and focus on getting the operator done and we could add examples later.

eyalfa committed:
akka-28364: address some more code review comments

eyalfa committed:
akka-28364: address some more review comments

eyalfa committed:
akka-28364: rename the thing

eyalfa committed:
akka-28364: rename impl class

eyalfa committed:
akka-28364: scalafmt

eyalfa committed:
Merge remote-tracking branch 'origin/master' into akka-28364

eyalfa committed:
akka-28364: updated copyrights statement.

johanandren commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFlatMapPrefixSpec.scala:
```diff
@@ -0,0 +1,534 @@
+
+  }
+
+}
```
A bunch of uncommented statement across this file, uncomment or remove

eyalfa replied:
if you're happy, I'm happy 😎

johanandren reviewed (commented):
Some tiny things more plus separate-file-it then I'm happy :)

eyalfa committed:
akka-28364: address final(?) review comments

eyalfa committed:
akka-28364: move to a separate file

eyalfa committed:
akka-28364: uodated copyright notice

eyalfa committed:
akka-28364: scalafmt

johanandren reviewed (approved):
LGTM, great work @eyalfa

jrudolph commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFlatMapPrefixSpec.scala:
```diff
@@ -0,0 +1,534 @@
+}
+import akka.{ Done, NotUsed }
+
+//import scala.concurrent.Future
```
left-over comments?

jrudolph commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFlatMapPrefixSpec.scala:
```diff
@@ -0,0 +1,534 @@
+          println("materializing flow")
+          Flow[Int]
+        }(Keep.left)
+        .alsoTo(Sink.foreach(println(_)))
```
Is that for debugging? Maybe we could remove it for now?

jrudolph commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFlatMapPrefixSpec.scala:
```diff
@@ -0,0 +1,534 @@
+        .fromPublisher(publisher)
+        .flatMapPrefixMat(2) { prefix =>
+          println("materializing flow")
+          Flow[Int].mapMaterializedValue(_ => prefix).prepend(Source(100 to 101)).alsoTo(Sink.foreach(println(_)))
```
Another `println` left-over?

jrudolph commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFlatMapPrefixSpec.scala:
```diff
@@ -0,0 +1,534 @@
+      suffixF.futureValue should ===(100 :: 101 :: Nil)
+    }
+
+    "handles upstream cancellation" in assertAllStagesStopped {
```
```suggestion
    "handles upstream completion" in assertAllStagesStopped {
```

jrudolph commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFlatMapPrefixSpec.scala:
```diff
@@ -0,0 +1,534 @@
+      val matValue = Source
+        .fromPublisher(publisher)
+        .flatMapPrefixMat(2) { prefix =>
+          println("materializing flow")
```
Remove `println`

jrudolph commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFlatMapPrefixSpec.scala:
```diff
@@ -0,0 +1,534 @@
+        .run
+
+      prefixF.futureValue should ===(0 until 4)
+      //Thread.sleep(1000 * 60 * 60)
```
Can be removed

jrudolph commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFlatMapPrefixSpec.scala:
```diff
@@ -0,0 +1,534 @@
+          println("materializing!")
+          prefix should ===(0 until 2)
+          Flow[Int].watchTermination()(Keep.right)
+        //Flow.fromSinkAndSource(Sink.fromSubscriber(subscriber), Source.fromPublisher(publisher))
```
Remove

jrudolph commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFlatMapPrefixSpec.scala:
```diff
@@ -0,0 +1,534 @@
+      subUpstream.sendNext(0)
+      //subUpstream.expectRequest() should be >= (1L)
+      subUpstream.sendNext(1)
+      //subUpstream.expectRequest() should be >= (1L)
```
Remove

jrudolph commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFlatMapPrefixSpec.scala:
```diff
@@ -0,0 +1,534 @@
+      subDownstream.request(1)
+      subUpstream.expectRequest() should be >= (1L)
+      subUpstream.sendNext(0)
+      //subUpstream.expectRequest() should be >= (1L)
```
Remove

jrudolph commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFlatMapPrefixSpec.scala:
```diff
@@ -0,0 +1,534 @@
+        .fromPublisher(publisher)
+        .watchTermination()(Keep.right)
+        .flatMapPrefixMat(8) { prefix =>
+          println("materializing!")
```
Remove

jrudolph commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFlatMapPrefixSpec.scala:
```diff
@@ -0,0 +1,534 @@
+      srcWatchTermF.futureValue should ===(Done)
+    }
+
+    "early downstream cancellation triggers flow materialization" in assertAllStagesStopped {
```
Is it explained somewhere why this would be desirable?

jrudolph commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFlatMapPrefixSpec.scala:
```diff
@@ -0,0 +1,534 @@
+      subUpstream.expectCancellation()
+    }
+
+    "early downstream failure triggers flow materialization" in assertAllStagesStopped {
```
I guess that could be folded together with the test before.

jrudolph commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFlatMapPrefixSpec.scala:
```diff
@@ -0,0 +1,534 @@
+        .watchTermination()(Keep.right)
+        .flatMapPrefixMat(8) { prefix =>
+          println("materializing!")
+          prefix should ===(0 until 2)
```
I wouldn't expect that :)

jrudolph commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFlatMapPrefixSpec.scala:
```diff
@@ -0,0 +1,534 @@
+
+  }
+
+}
```
Very nice coverage!

jrudolph commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FlatMapPrefix.scala:
```diff
@@ -0,0 +1,156 @@
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[M]) = {
+    val matPromise = Promise[M]
+    val logic = new GraphStageLogic(shape) with InHandler with OutHandler {
+      var accumulated = List.empty[In]
```
I'd propose to use a `VectorBuilder` with `sizeHint(n)` instead.

jrudolph replied:
Mostly, because `List.size` is O(n).

jrudolph commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FlatMapPrefix.scala:
```diff
@@ -0,0 +1,156 @@
+          subSource.get.push(grab(in))
+        } else {
+          accumulated ::= grab(in)
+          if (accumulated.size == n) {
```
`List.size` is O(n) => use VectorBuilder :)

eyalfa replied:
notice `accumulated' is now a `scala.collection.mutable.Buffer'

jrudolph commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FlatMapPrefix.scala:
```diff
@@ -0,0 +1,156 @@
+            matPromise.failure(new NeverMaterializedException(ex))
+            subSource = OptionVal.None
+            subSink = OptionVal.None
+            throw ex
```
Use failStage instead to prevent extra log messages (plus points for adding an `EventFilter` to the test to check that no extra log message was thrown).

eyalfa replied:
can u please point me to an example of using such an event filter in tests?

eyalfa replied:
re. failStage, I need this exception to terminate this method's execution. this is easily achieved by throwing an exception, reverting to failStag will require introduction of a boolean var - which I'd be happy to avoid if possible

jrudolph replied:
Then there needs to be a try/catch around materializeFlow() which ultimately calls `failStage`.

jrudolph commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FlatMapPrefix.scala:
```diff
@@ -0,0 +1,156 @@
+import scala.concurrent.{ Future, Promise }
+import scala.util.control.NonFatal
+
+@InternalApi private[akka] final class FlatMapPrefix[In, Out, M](n: Int, f: immutable.Seq[In] => Flow[In, Out, M])
```
Didn't check the implementation for correctness but trusted the tests.

eyalfa replied:
is this comment still relevant? :-)

jrudolph commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FlatMapPrefix.scala:
```diff
@@ -0,0 +1,156 @@
+      override def postStop(): Unit = {
+        matPromise.tryFailure(new AbruptStageTerminationException(this))
+        if (subSource.isDefined && !subSource.get.isClosed)
+          subSource.get.complete()
```
Should that source be failed there?

eyalfa replied:
hmm... this is abrubt-termination land, we might be able to rely on the interpreter calling post-stop on the materialized flow as well. this way we'd get the materialized flow's behaviour: either failing or completing, I'm just not sure if we'd get the future matval properly completed...
will have to look a bit deeper into this

eyalfa replied:
thinking about this a bit further:
substop can be calles in one of two cases: 
1. after stage completion, hence all ports (in and out) were closed.
2. abrupt termination.

case 1 implies that onUpstreamCompletion/failure and onDownstreamCompletion(failure) were invoked, hence we had our chance of notifying the materialized flow of this events, hence in this case we should **not** close the ports during postStop.
case 2 implies that the materialized flow will soon be stopped as well (if not already stopped), hence again - no need to close its ports.

@johanandren , @jrudolph , does this make any sense to you guys?

johanandren replied:
The subflow will always be same materializer of this graphstage so I think it'd be safe to rely on that abruptly terminating the nested flow as well.

jrudolph commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FlatMapPrefix.scala:
```diff
@@ -0,0 +1,156 @@
+        if (subSource.isDefined && !subSource.get.isClosed)
+          subSource.get.complete()
+        if (subSink.isDefined && !subSink.get.isClosed)
+          subSink.get.cancel()
```
Add `AbruptStageTerminationException` as cause?

eyalfa replied:
same question here, can we rely on the interpreter to call postStop on the materialized flow?

jrudolph commented on akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala:
```diff
@@ -9,15 +9,11 @@ import java.util.concurrent.atomic.AtomicReference
 import akka.annotation.InternalApi
-import akka.stream.ActorAttributes.StreamSubscriptionTimeout
-import akka.stream.ActorAttributes.SupervisionStrategy
+import akka.stream.ActorAttributes.{ StreamSubscriptionTimeout, SupervisionStrategy }
```
Nitpick: do we actually like that style more? Otherwise, better leave unchanged source files alone to avoid conflicts.

johanandren replied:
We try to avoid these multi-imports since they easily cause git conflicts.

johanandren replied:
(I don't think that should block merging this PR though)

eyalfa replied:
I think this is the result of idea's auto-imports, I'll revert it

jrudolph commented on akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala:
```diff
@@ -2024,6 +2024,47 @@ final class Flow[In, Out, Mat](delegate: scaladsl.Flow[In, Out, Mat]) extends Gr
 
+  /**
+   * Takes up to `n` elements from the stream (less than `n` only if the upstream completes before emitting `n` elements),
+   * then apply f on these elements in order to obtain a flow, this flow is then materialized and the rest of the input is processed by this flow (similar to via ).
```
```suggestion
   * then apply `f` on these elements in order to obtain a flow, this flow is then materialized and the rest of the input is processed by this flow (similar to `via`).
```

jrudolph commented on akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala:
```diff
@@ -2024,6 +2024,47 @@ final class Flow[In, Out, Mat](delegate: scaladsl.Flow[In, Out, Mat]) extends Gr
+   * This method returns a flow consuming the rest of the stream producing the materialized flow's output.
+   *
+   * '''Emits when''' the materialized flow emits.
+   *  Notice the first `n` elements are buffered internally before materializing the flow, This flow will then be materialized and connected to the rest of the upstream - producing elements at its own discretion (might 'swallow' or multiply elements).
```
```suggestion
   *  Notice the first `n` elements are buffered internally before materializing the flow and connecting it to the rest of the upstream - producing elements at its own discretion (might 'swallow' or multiply elements).
```

(Not sure if the notice is needed that the new flow might have arbitrary semantics, that seems obvious to me and the same as for `via`)

jrudolph commented on akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala:
```diff
@@ -2024,6 +2024,47 @@ final class Flow[In, Out, Mat](delegate: scaladsl.Flow[In, Out, Mat]) extends Gr
+   *
+   * '''Completes when''' the materialized flow completes.
+   *  If upstream completes before producing `n` elements, `f` will be applied with the provided elements,
+   *  the resulting flow will be materialized and signalled for upstream completion, it can then cancel/complete at its own discretion.
```
```suggestion
   *  the resulting flow will be materialized and signalled for upstream completion, it can then complete or continue to emit elements at its own discretion.
```

jrudolph commented on akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala:
```diff
@@ -2024,6 +2024,47 @@ final class Flow[In, Out, Mat](delegate: scaladsl.Flow[In, Out, Mat]) extends Gr
+   *  If upstream completes before producing `n` elements, `f` will be applied with the provided elements,
+   *  the resulting flow will be materialized and signalled for upstream completion, it can then cancel/complete at its own discretion.
+   *
+   * '''Cancels when''' upstream cancels, application of `f` fails, materialization of the resulting flow fails or the materialized flow cancels.
```
```suggestion
   * '''Cancels when''' downstream cancels before the prefix was collected, application of `f` fails, materialization of the resulting flow fails or the materialized flow cancels.
```

eyalfa replied:
@jrudolph , this sentence will become obsolete once I implement your major comment about downstream cancellation prior to prefix completion.
I guess it should say something like:

'''Cancels when''' the materialized flow cancels or when failing to materialize the flow.

we also need to add a description of the way downstream cancellation is handled before the prefix is 'ready' (we've agreed on buffering the cause and feeding it to the materialized flow once materialized)/
should this description be part of the 'cancels when' section or a separate section?

jrudolph replied:
Let's put it all into this section. In any case, it still needs to be fixed.

jrudolph commented on akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala:
```diff
@@ -2024,6 +2024,47 @@ final class Flow[In, Out, Mat](delegate: scaladsl.Flow[In, Out, Mat]) extends Gr
+   *  the resulting flow will be materialized and signalled for upstream completion, it can then cancel/complete at its own discretion.
+   *
+   * '''Cancels when''' upstream cancels, application of `f` fails, materialization of the resulting flow fails or the materialized flow cancels.
+   *  Notice the materialized flow may apply its own logic in case of upstream or downstream cancellations.
```
```suggestion
   *  Notice that the materialized flow may apply its own logic in case of upstream or downstream cancellations.
```

jrudolph commented on akka-stream/src/main/scala/akka/stream/javadsl/Source.scala:
```diff
@@ -3148,6 +3148,47 @@ final class Source[Out, Mat](delegate: scaladsl.Source[Out, Mat]) extends Graph[
     new Source(delegate.prefixAndTail(n).map { case (taken, tail) => Pair(taken.asJava, tail.asJava) })
 
+  /**
+   * Takes up to `n` elements from the stream (less than `n` only if the upstream completes before emitting `n` elements),
```
Propagate applied changes from the scaladsl version.

jrudolph commented on akka-stream/src/main/scala/akka/stream/javadsl/SubFlow.scala:
```diff
@@ -1380,6 +1380,33 @@ class SubFlow[In, Out, Mat](
     Mat] =
     new SubFlow(delegate.prefixAndTail(n).map { case (taken, tail) => akka.japi.Pair(taken.asJava, tail.asJava) })
 
+  /**
```
Propagate changes

jrudolph commented on akka-stream/src/main/scala/akka/stream/javadsl/SubSource.scala:
```diff
@@ -1357,6 +1357,33 @@ class SubSource[Out, Mat](
     Mat] =
     new SubSource(delegate.prefixAndTail(n).map { case (taken, tail) => akka.japi.Pair(taken.asJava, tail.asJava) })
 
+  /**
```
Propagate changes

jrudolph commented on akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:
```diff
@@ -1929,6 +1929,30 @@ trait FlowOps[+Out, +Mat] {
   def prefixAndTail[U >: Out](n: Int): Repr[(immutable.Seq[Out], Source[U, NotUsed])] =
     via(new PrefixAndTail[Out](n))
 
+  /**
```
Propagate changes

jrudolph commented on akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala:
```diff
@@ -2024,6 +2024,47 @@ final class Flow[In, Out, Mat](delegate: scaladsl.Flow[In, Out, Mat]) extends Gr
+
+  /**
+   * mat version of [[#flatMapPrefix]], this method gives access to a future materialized value of the downstream flow (as a completion stage).
+   *see [[#flatMapPrefix]] for details.
```
```suggestion
   * see [[#flatMapPrefix]] for details.
```

jrudolph commented on akka-docs/src/main/paradox/stream/operators/Source-or-Flow/flatMapPrefix.md:
```diff
@@ -0,0 +1,37 @@
+# flatMapPrefix
```
I proposed a few changes to the scaladoc of `flatMapPrefix`, these should probably be propagated here as well where applicable.

jrudolph reviewed (commented):
Great work. Some cosmetic changes and two main issues:

 * Could you explain (or point me to) the reasoning for the behavior in the test "early downstream cancellation triggers flow materialization"? I don't find it intuitive or desirable.
 * Use something else than `List` for buffering elements in the stage (`List.size` is O(n), leading to O(n^2) behavior which might show for big prefixes)

eyalfa commented:
@jrudolph , thank you very much for the swift review.
I'll try addressing most of your comments and suggestions tonight,
re. you question about '"early downstream cancellation triggers flow materialization"':
while most flows will terminate on downstream cancellation some might keep running, few examples are Flow.fromSinkAndSource, or a graph that goes through a broadcast with eagerCancel=false. I believe these flows should exhibit the same behaviour when wrapped with this operator, I guess the best way to look at it is simulating `flatMapPrefix(0)(...)` and making sure it behaves exactly like the materialized flow (as if it was used directly and not under this operator). I think this view also holds when `n>0`, since once the materialized flow takes charge it should behave exactly the same as if it was fed with the upstream's tail.

jrudolph commented:
> re. you question about '"early downstream cancellation triggers flow materialization"':
> while most flows will terminate on downstream cancellation some might keep running, few examples are Flow.fromSinkAndSource, or a graph that goes through a broadcast with eagerCancel=false. I believe these flows should exhibit the same behaviour when wrapped with this operator, I guess the best way to look at it is simulating `flatMapPrefix(0)(...)` and making sure it behaves exactly like the materialized flow (as if it was used directly and not under this operator). I think this view also holds when `n>0`, since once the materialized flow takes charge it should behave exactly the same as if it was fed with the upstream's tail.

I think that behavior makes sense only after the prefix was actually received. In the test, you have

```
.flatMapPrefixMat(8) { prefix =>
          prefix should ===(0 until 2)
```

when the result stream is cancelled after two elements. That doesn't make sense to me because the inner flow is misled to assume that the original stream only had two elements when indeed it had more but was just cancelled too early.

Do you have any opinion, @johanandren?

johanandren commented:
Hmmm, that's a good point. A bit on the fence because one hand it is kind of weird that the factory is called with a shorter prefix, but on the other hand I think it would make the following stream which keeps consuming regardless of early downstream cancelation:

```
Source.whatever
  .flatMapPrefix(8)(prefix => 
    Flow.fromSinkAndSource(Sink.consumeForever, Source.whatever)
  ).to(Sink.cancelsQuickly)
```

Perhaps that is not a problem and just made up use case though.

Or perhaps it should not cancel the stage, but also not materialize inner flow early?

jrudolph commented:
> Hmmm, that's a good point. A bit on the fence because one hand it is kind of weird that the factory is called with a shorter prefix, but on the other hand I think it would make the following stream which keeps consuming regardless of early downstream cancelation:
> 
> ```
> Source.whatever
>   .flatMapPrefix(8)(prefix => 
>     Flow.fromSinkAndSource(Sink.consumeForever, Source.whatever)
>   ).to(Sink.cancelsQuickly)
> ```

Ah, that's an interesting example. In fact, I would assume that downstream demand should have no influence whatever over the original stream (only if the later materialized flow propagates it) but in the test downstream cancellation actually cuts off the prefix collection.

I think there should be tests for:
  * collection of prefix is decoupled from downstream demand
  * in particular, downstream cancellation needs to be buffered until the flow can be materialized and should then be passed to the flow.
  * after flow has been materialized, downstream demand might be propagated or not to the original stream depending on the materialized flow

(I haven't looked at whether part of those tests is already available)

eyalfa commented:
@jrudolph , @johanandren ,
The more I think of it the more I understand @jrudolph claim (which probably means more work for me 😎 )

long story short, we have few options here:
1. store the downstream cancellation reason, keep consuming upstream, materialize the flow and hand it the cancellation reason. probably requires some observation of the cause in order to identify severe exceptions (or is the interpreter guaranteed to stop all stages in such cases?)
2. cancel/fail the stage without materializing the flow, this is a very simple approach and one which is easily explained to users of the library. advanced users having 'crazy' use-cases like the one shown in the test (flow keeps running though its downstream lost interest) can intercept downstream cancellation/failures using existing/customised stages between the flatMapPrefix stage and its downstream. this has the benefit of simplicity + one class one responsibility.
3. document the current behaviour as is, extreme use-cases can be handled as described in #2.
4. parameterise this method with a policy... I think this'll loose potential users before even reading the docs...

I think we should probably consider options 1 & 2, 1 is more 'complete' 2 is simpler.

WDUT guys?

johanandren commented:
I like option 1 better. You can decide if it is fatal using pattern matching with `scala.util.control.NonFatal` (but I wonder if a fatal error would ever propagate as a cancellation reason)

eyalfa committed:
akka-28364: address most comments by @jrodulph

eyalfa committed:
akka-28364: scalafmtAll

eyalfa committed:
akka-28364: fix/improve postStop handling, add another test for abrupt termination.

eyalfa committed:
akka-28364: add an explicit counter as part of accumulated, avoid o(n) size invocations.

eyalfa committed:
akka-28364: modify one of the tests to reflect the new behaviour when downstream completes before prefix is complete.

eyalfa committed:
akka-28364: modify the test case for early downstream failure.

eyalfa committed:
akka-28364: fix the stage's implementation to buffer early downstream errors.

eyalfa committed:
akka-28364: scalafmtAll

eyalfa commented:
@johanandren , @jrudolph , I've fixed the latest comments especially those related to early downstream completion and postStop behavior.
I'd appreciate your review.

johanandren reviewed (approved):
LGTM but still some smaller things in comments by Johannes that were not addressed.

jrudolph commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFlatMapPrefixSpec.scala:
```diff
@@ -0,0 +1,553 @@
+import akka.{ Done, NotUsed }
+
+class FlowFlatMapPrefixSpec extends StreamSpec {
+  //import system.dispatcher
```
Here and further down are more comments that should be removed.

jrudolph commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFlatMapPrefixSpec.scala:
```diff
@@ -321,22 +313,29 @@ class FlowFlatMapPrefixSpec extends StreamSpec {
+        .flatMapPrefixMat(3) { prefix =>
+          prefix should ===(0 until 3)
           Flow[Int].watchTermination()(Keep.right)
         //Flow.fromSinkAndSource(Sink.fromSubscriber(subscriber), Source.fromPublisher(publisher))
```
more comments

jrudolph commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFlatMapPrefixSpec.scala:
```diff
@@ -357,6 +356,11 @@ class FlowFlatMapPrefixSpec extends StreamSpec {
       //subUpstream.expectRequest() should be >= (1L)
```
here as well

jrudolph commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFlatMapPrefixSpec.scala:
```diff
@@ -321,22 +313,29 @@ class FlowFlatMapPrefixSpec extends StreamSpec {
+      //this one is sent AFTER downstream cancellation
+      subUpstream.sendNext(2)
+
+      subUpstream.expectCancellation()
```
That looks better now! :+1:

jrudolph commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FlatMapPrefix.scala:
```diff
@@ -26,28 +26,28 @@ import scala.util.control.NonFatal
     val matPromise = Promise[M]
     val logic = new GraphStageLogic(shape) with InHandler with OutHandler {
-      var accumulated = List.empty[In]
+      var accumulated = (List.empty[In], 0)
```
I'd still appreciate using a buffer here as suggested before. That will use less allocations and will make the code better readable.

jrudolph commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FlatMapPrefix.scala:
```diff
@@ -78,23 +78,29 @@ import scala.util.control.NonFatal
+            case NonFatal(ex) =>
+              downstreamCause = OptionVal.Some(ex)
+            case ex =>
+              super.onDownstreamFinish(ex)
```
I think that's not needed. Fatal errors won't be propagated (if they are it's a bug).

jrudolph reviewed (commented):
Still a few more things left.

eyalfa committed:
akka-28364: address @jrudolph comments

eyalfa committed:
akka-28364: address some more comments from @jrudolph.

eyalfa committed:
akka-28364: scalafmt

eyalfa commented:
@jrudolph , I've addressed all of your comments (though for some reason github didn't mark all of them 'outdated'), can you please conuct one more (final?) review?

patriknw reviewed (approved):
LGTM

patriknw commented:
@johanandren Do you think Johannes has reviewed this thoroughly enough so that we can merge it?

patriknw commented:
Thanks for great contribution @eyalfa

eyalfa commented:
@patriknw thanks for your review!
what's next, wait for @jrudolph final review?

johanandren commented:
I think it would be better to get a 👍 from Johannes than eager-merge this (even if it would be nice to get it in the next release).

eyalfa commented:
@jrudolph , any chance to finalize this?

jrudolph commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FlatMapPrefix.scala:
```diff
@@ -0,0 +1,165 @@
+                if (!isClosed(in)) {
+                  cancel(in, cause)
+                }
+                //super.onDownstreamFinish(cause)
```
Extra comment

jrudolph commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FlatMapPrefix.scala:
```diff
@@ -0,0 +1,165 @@
+        try {
+          val prefix = accumulated.toVector
+          accumulated.clear()
+          subSource = OptionVal.Some(new SubSourceOutlet[In]("subSource"))
```
Also add class name to name so it can be more easily identified.

jrudolph commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FlatMapPrefix.scala:
```diff
@@ -0,0 +1,165 @@
+          subSink.get.setHandler {
+            new InHandler {
+              override def onPush(): Unit = {
+                push(out, subSink.get.grab())
```
This is in the fast path, so maybe get rid of the `.get` which introduces an extra null check. If you do that also get rid of all the other `.get` in here.

jrudolph replied:
This particular one hasn't been fixed yet ;)

jrudolph reviewed (commented):
Sorry, have been traveling etc in the last week.

Some final nitpicks but otherwise LGTM :)

eyalfa committed:
akka-28364: address @jrudolph comments.

eyalfa committed:
akka-28364: scalafmt

jrudolph commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FlatMapPrefix.scala:
```diff
@@ -42,62 +45,62 @@ import scala.util.control.NonFatal
-          subSource = OptionVal.Some(new SubSourceOutlet[In]("subSource"))
-          subSource.get.setHandler {
+          subSource = OptionVal.Some(new SubSourceOutlet[In](s"${this}.subSource"))
+          subSource.x.setHandler {
```
That's probably fast enough but maybe just let's create another `val` with the value and just use that one in here?

Like 

```scala
val theSubSource = new SubSourceOutlet[In](s"${this}.subSource")
this.subSource = OptionVal.Some(theSubSource)
theSubSource.setHandler
...
```

jrudolph commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FlatMapPrefix.scala:
```diff
@@ -42,62 +45,62 @@ import scala.util.control.NonFatal
-            //gi'me some more!
-            pull(in)
-          }
+        subSource match {
```
:+1: Good to use matching here and below.

jrudolph reviewed (commented):
Another smallish comment

eyalfa committed:
akka-28364: address last(?) comments from @jrudolph

jrudolph commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FlatMapPrefix.scala:
```diff
@@ -0,0 +1,169 @@
+
+          //in case downstream was closed
+          if (downstreamCause.isDefined) {
+            subSink.x.cancel(downstreamCause.get)
```
Here

jrudolph commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FlatMapPrefix.scala:
```diff
@@ -0,0 +1,169 @@
+
+          //in case we've been pulled by downstream
+          if (isAvailable(out)) {
+            subSink.x.pull()
```
here

jrudolph commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FlatMapPrefix.scala:
```diff
@@ -0,0 +1,169 @@
+
+          //in case we've materialized due to upstream completion
+          if (isClosed(in)) {
+            subSource.x.complete()
```
here

jrudolph commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FlatMapPrefix.scala:
```diff
@@ -0,0 +1,169 @@
+          matPromise.success(matVal)
+
+          //in case downstream was closed
+          if (downstreamCause.isDefined) {
```
Maybe use `downstreamCause.foreach` (if that's defined on OptionVal, didn't check)?

jrudolph commented on akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala:
```diff
@@ -2024,6 +2024,47 @@ final class Flow[In, Out, Mat](delegate: scaladsl.Flow[In, Out, Mat]) extends Gr
+   * '''Emits when''' the materialized flow emits.
+   *  Notice the first `n` elements are buffered internally before materializing the flow and connecting it to the rest of the upstream - producing elements at its own discretion (might 'swallow' or multiply elements).
+   *
+   * '''Backpressures when''' downstream backpressures
```
Is that still right, though? Maybe something like

```suggestion
   * '''Backpressures''' after the prefix has been collected when the materialized flow backpressures. The prefix is collected eagerly without applying backpressure.
```

That then needs to be propagated. You could also do that change in a follow-up PR, so that this one can be already merged.

eyalfa replied:
this stage actually doesn't back pressures until materializing the flow,  it will keep consuming as long as it have to fill the prefix (very similar to buffer in this respect).
it will only start backpressuring  once the flow is materialized and backpressures on its own.

so basically I think the wording is right, it might make sense to add some more wording in the docs about the detailed behaviour of this operator.

jrudolph replied:
Currently it says

> '''Backpressures when''' downstream backpressures

which doesn't seem correct.

jrudolph reviewed (commented):
Sorry to be so picky but a few instances were left over :)

eyalfa committed:
akka-28364: one more round of review comments.

eyalfa committed:
akka-28364: scalafmt.

jrudolph reviewed (approved):
LGTM, created follow up https://github.com/akka/akka/issues/28574 to fix the slight doc issue I commented on before so we can merge this now.

eyalfa commented:
@jrudolph , can u please merge?

jrudolph commented:
Thanks a lot, @eyalfa. :tada:
